Help Center: block access to chat from the get go when ZD is blocked

alshakero opened the pull request:
Fixes DOTCOM-13797

## Proposed Changes

* Enhanced the `useChatStatus` hook to check if users can connect to Zendesk messaging before determining chat eligibility
* This way, users without ZD access are considered illegible and are never stuck in the limbo between Odie and a broken chat.

## Why are these changes being made?
Users are currently stranded between Odie and support upon escalation.

## Testing Instructions

1. Open the live link.
2. Open DevTools > Network
3. Filter by embeddable/config.
4. Right-click on this request and block its URL.
5. Refresh the page.
6. Open the Help Center.
7. You shouldn't be able to access chat. 
8. Ask Odie for a human,
9. It should say (which is inaccurate but separate)
![image](https://github.com/user-attachments/assets/4fb74a71-ac19-41f8-a1b6-2a93fc4227ba)

alshakero committed:
Block access to chat from the get go when ZD is blocked

xavier-lc commented:
> 7\. You shouldn't be able to access chat.

I can access it:

![image](https://github.com/user-attachments/assets/980867cd-a35e-411d-8b00-91e71309e94b)

alshakero committed:
Merge branch 'trunk' into dotcom-13797-help-center-fail-gracefully-when-users-cant-reach-support

alshakero committed:
Ensure a boolean

alshakero committed:
Merge branch 'dotcom-13797-help-center-fail-gracefully-when-users-cant-reach-support' of github.com:Automattic/wp-calypso into dotcom-13797-help-center-fail-gracefully-when-users-cant-reach-support

alshakero committed:
Merge branch 'trunk' into dotcom-13797-help-center-fail-gracefully-when-users-cant-reach-support

alshakero committed:
Address feedback

alshakero commented:
@rcrdortiz @xavier-lc thank you gentlemen, I address all the feedback and tested.

xavier-lc commented on packages/help-center/src/hooks/use-chat-status.ts:
```diff
@@ -1,14 +1,19 @@
 	// See: pdDR7T-1vN-p2
-	const isEligibleForChat = Boolean( supportStatus?.eligibility?.is_user_eligible );
+	// They're only eligible if they can connect to Zendesk.
+	const isEligibleForChat = Boolean(
+		supportStatus?.eligibility?.is_user_eligible && canConnectToZendesk
+	);
```
I find it confusing that this `isEligibleForChat` condition is different from the one on the file avobe.

xavier-lc replied:
When `canConnectToZendesk` is `false`, `supportStatus?.eligibility?.is_user_eligible` will be `null`, right? So maybe we can skip it.

alshakero replied:
> I find it confusing that this isEligibleForChat condition is different from the one on the file avobe.

Yes `isEligibleForChat` became blunt (not 100% sure this is a good idea). It now encapsulates both pieces of information (`is_user_eligible` and `canConnectToZendesk`). Well, we need to know both of these in `notices` because we need to know if the user is `is_user_eligible` in separation of their ZD connectivity status.

> When canConnectToZendesk is false, supportStatus?.eligibility?.is_user_eligible will be null, right? So maybe we can skip it.

This is too confusing, I tried imagining the truth table, and it's too complicated. I'll rethink this.

xavier-lc commented on packages/help-center/src/hooks/use-chat-status.ts:
```diff
@@ -1,14 +1,19 @@
 
 export default function useChatStatus() {
-	const { data: supportStatus } = useSupportStatus();
+	const { data: canConnectToZendesk } = useCanConnectToZendeskMessaging();
```
Also confused by the fact that `canConnectToZendesk` is returned by `useHelpCenterContext`. Is it the same value?

alshakero replied:
Yes, that's a mistake in the design that I didn't want to refactor here. It's a common mistake where people pass down or contexualize values that are globally accessible via hooks.

alshakero replied:
Removed the contextualized canConnectToZendesk.

xavier-lc reviewed (commented):
Works as described, left some comments.

alshakero committed:
remove `canConnectToZendesk` from context

alshakero committed:
Ensure `canConnectToZendesk` for every instance of `isEligibleForChat`

alshakero committed:
Leave `useChatStatus` alone

alshakero committed:
Merge branch 'trunk' into dotcom-13797-help-center-fail-gracefully-when-users-cant-reach-support

alshakero committed:
Change hook import

alshakero commented:
@xavier-lc OK this is good to go. I left the useChatStatus hook untouched and added an accompanying check to `isEligibleForChat` everywhere it is used. This way it's much clearer, albeit slightly more verbose.

I also removed the duplicated contextualized values.

alshakero commented on packages/help-center/src/components/help-center-chat.tsx:
```diff
@@ -46,7 +47,7 @@ export function HelpCenterChat( {
 		<OdieAssistantProvider
 			currentUser={ currentUser }
 			canConnectToZendesk={ canConnectToZendesk }
-			isLoadingCanConnectToZendesk={ isLoadingCanConnectToZendesk }
+			isLoadingCanConnectToZendesk={ isLoading }
```
This contextualized value can also be removed but I didn't want to be overzealous.

xavier-lc commented on packages/odie-client/src/data/use-get-support-interactions.ts:
```diff
@@ -17,9 +17,20 @@ export const useGetSupportInteractions = (
 	const isTestMode = isTestModeEnvironment();
+	const { data: canConnectToZendesk } = useCanConnectToZendeskMessaging( enabled );
+	// Only fetch Zendesk interactions if the user can connect to Zendesk.
+	const shouldFetch = enabled && ( canConnectToZendesk || ! provider?.includes( 'zendesk' ) );
```
Could you explain the `! provider?.includes( 'zendesk' )` bit, please?

alshakero replied:
Yes, sometimes it's zendesk, sometimes it's zendesk-staging.

xavier-lc replied:
But why do we want the `provider` to not be `zendesk`?

alshakero replied:
When enabled is true:

<ul>
<li>
<p>shouldFetch will be <strong>false</strong> whenever <code inline="">enabled</code> is <code inline="">false</code>.</p>
</li>
<li>
<p>When <code inline="">enabled</code> is <code inline="">true</code>:</p>
<ul>
<li>
<p>If <code inline="">canConnectToZendesk</code> is <code inline="">true</code>, the result is <strong>true</strong> regardless of the provider.</p>
</li>
<li>
<p>If <code inline="">canConnectToZendesk</code> is <code inline="">false</code>, the result is <strong>true</strong> only when the provider string <strong>does not</strong> contain “zendesk” (i.e., <code inline="">odie</code>).</p>
</li>
</ul>
</li>
</ul></body></html>

mmtr replied:
I think the comment is a bit misleading then, because we fetch Zendesk interactions even if the user cannot connect to Zendesk.

mmtr replied:
Oooh, it's because this fetches interactions filtered by the provider, so it's not actually fetching Zendesk interactions in these circumstances.

So, if `canConnectToZendesk` is false, this will fetch interactions from other providers (`odie` and `help-center`).

xavier-lc replied:
I agree with Miguel.

xavier-lc commented on packages/odie-client/src/data/use-get-support-interactions.ts:
```diff
@@ -31,11 +42,18 @@ export const useGetSupportInteractions = (
 				return response.filter( ( interaction ) =>
 					interaction.events.some( ( event ) => event.event_source === provider )
 				);
+			} else if ( ! canConnectToZendesk ) {
+				// When provider is null, we're fetching interactions from all providers.
+				// We need to filter out Zendesk interactions if the user can't connect to Zendesk.
+				return response.filter(
+					( interaction ) =>
+						! interaction.events.some( ( event ) => event.event_source.includes( 'zendesk' ) )
+				);
```
If both `provider` and `canConnectToZendesk` are falsy, then we return the `response` without any filtering, right?

alshakero replied:
If `provider` is falsy, it means we'll get interactions from all providers, including Zendesk.

mmtr replied:
> If both `provider` and `canConnectToZendesk` are falsy, then we return the response without any filtering, right?

I don't think so, according to the code. When both are falsy, we return interactions from all providers except Zendesk.

alshakero replied:
I meant to say "**Unless I add this condition**, if the provider is `falsy`, it means we'll get interactions from all providers, including Zendesk."

xavier-lc reviewed (commented):
Excellent, just a couple questions and I'm giving the ✅, I promise 😅

mmtr commented on packages/odie-client/src/data/use-get-support-interactions.ts:
```diff
@@ -17,9 +17,20 @@ export const useGetSupportInteractions = (
 	const path = `?per_page=${ per_page }&page=${ page }&status=${ status }`;
 	const isTestMode = isTestModeEnvironment();
+	const { data: canConnectToZendesk } = useCanConnectToZendeskMessaging( enabled );
+	// Only fetch Zendesk interactions if the user can connect to Zendesk.
+	const shouldFetch = enabled && ( canConnectToZendesk || ! provider?.includes( 'zendesk' ) );
```
As per discussion above, this can lead to assume that the code can only fetch Zendesk interactions, if the code is not read very carefully.

I'd slightly change it to make it clearer (albeit more verbose):
```suggestion
	// Only fetch interactions if the user can connect to the provider.
	let shouldFetch = enabled;
	if ( provider?.includes( 'zendesk' ) && ! canConnectToZendesk ) {
		shouldFetch = false;
	}
```

alshakero replied:
This is perfect. Thanks!

mmtr reviewed (commented):
This tests well for me too. Just left a minor note about code legibility.

alshakero committed:
Improve code eligibility

alshakero commented:
> Excellent, just a couple questions and I'm giving the ✅, I promise 😅

I don't mind at all! I absolutely love a demanding review. Don't hold back and reject as many times as you want :D
